client/src/Pos/Client/KeyStorage.hs

```diff
@@ -3,120 +3,139 @@
+    , modifySecretPureDefault
+    , modifyPublicPureDefault
+    , modifySecretDefault
+    , modifyPublicDefault
+    , getSecretKeys
+    , getPublicKeys
+    , getSecretKeysPlain
+    , getPublicKeysPlain
+    , addSecretKey
+    , addPublicKey
+    , deleteAllSecretKeys
+    , deleteAllPublicKeys
+    , deleteSecretKeyBy
+    , deletePublicKeyBy
+    ) where
 
 import           Universum
 
 import qualified Control.Concurrent.STM as STM
 import           Control.Lens ((<%=), (<>~))
 import           Serokell.Util (modifyTVarS)
-import           System.Wlog (WithLogger)
 
 import           Pos.Binary.Crypto ()
-import           Pos.Crypto (EncryptedSecretKey, PassPhrase, SecretKey, hash, runSecureRandom,
-                             safeKeyGen)
-import           Pos.Util.UserSecret (HasUserSecret (..), UserSecret, peekUserSecret, usKeys,
-                                      usPrimKey, writeUserSecret)
-
-type KeyData = TVar UserSecret
+import           Pos.Crypto (EncryptedSecretKey, PublicKey, hash)
+import           Pos.Util.UserSecret (HasUserSecret (..), UserSecret, usKeys,
+                                      writeUserSecret)
+import           Pos.Util.UserPublic (HasUserPublic (..), UserPublic, upKeys,
+                                      writeUserPublic)
 
 ----------------------------------------------------------------------
 -- MonadKeys class and default functions
 ----------------------------------------------------------------------
 
 class Monad m => MonadKeysRead m where
     getSecret :: m UserSecret
+    getPublic :: m UserPublic
 
 class MonadKeysRead m => MonadKeys m where
     modifySecret :: (UserSecret -> UserSecret) -> m ()
+    modifyPublic :: (UserPublic -> UserPublic) -> m ()
 
 type HasKeysContext ctx m =
     ( MonadReader ctx m
     , HasUserSecret ctx
+    , HasUserPublic ctx
     , MonadIO m
     )
 
 getSecretDefault :: HasKeysContext ctx m => m UserSecret
 getSecretDefault = view userSecret >>= atomically . STM.readTVar
 
+getPublicDefault :: HasKeysContext ctx m => m UserPublic
+getPublicDefault = view userPublic >>= atomically . STM.readTVar
+
 modifySecretPureDefault :: HasKeysContext ctx m => (UserSecret -> UserSecret) -> m ()
 modifySecretPureDefault f = do
     us <- view userSecret
     void $ atomically $ modifyTVarS us (identity <%= f)
 
+modifyPublicPureDefault :: HasKeysContext ctx m => (UserPublic -> UserPublic) -> m ()
+modifyPublicPureDefault f = do
+    up <- view userPublic
+    void $ atomically $ modifyTVarS up (identity <%= f)
+
 modifySecretDefault :: HasKeysContext ctx m => (UserSecret -> UserSecret) -> m ()
 modifySecretDefault f = do
     us <- view userSecret
     new <- atomically $ modifyTVarS us (identity <%= f)
     writeUserSecret new
 
+modifyPublicDefault :: HasKeysContext ctx m => (UserPublic -> UserPublic) -> m ()
+modifyPublicDefault f = do
+    up <- view userPublic
+    new <- atomically $ modifyTVarS up (identity <%= f)
+    writeUserPublic new
```

Comment:
Function `writeUserPublic` is the place where public key is writing to the local file.